app/models/internal/impl/PagerDutyNotificationEntry.java

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright 2018 Smartsheet.com
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package models.internal.impl;
+
+import akka.actor.ActorSystem;
+import akka.http.javadsl.Http;
+import akka.http.javadsl.model.ContentTypes;
+import akka.http.javadsl.model.HttpEntities;
+import akka.http.javadsl.model.HttpRequest;
+import com.arpnetworking.commons.builder.OvalBuilder;
+import com.arpnetworking.mql.grammar.AlertTrigger;
+import com.arpnetworking.steno.Logger;
+import com.arpnetworking.steno.LoggerFactory;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.google.inject.Injector;
+import com.typesafe.config.Config;
+import models.internal.Alert;
+import models.internal.NotificationEntry;
+
+import java.net.URI;
+import java.util.Objects;
+import java.util.concurrent.CompletionStage;
+import java.util.stream.Collectors;
+
+/**
+ * Internal model representing a pagerduty notification entry.
+ * These do not have any internal state and act as singletons when creating a NotificationGroup.
+ *
+ * @author Sheldon White (sheldon.white at smartsheet dot com)
+ */
+public final class PagerDutyNotificationEntry implements NotificationEntry {
+    @Override
+    public CompletionStage<Void> notifyRecipient(final Alert alert, final AlertTrigger trigger, final Injector injector) {
+        LOGGER.debug().setMessage("executing pagerduty call").log();
+        final Config typesafeConfig = injector.getInstance(Config.class);
+        String pagerDutyEndpoint = typesafeConfig.getString("pagerDuty.uri");
+        String pagerDutyServiceKey = typesafeConfig.getString("pagerDuty.serviceKey");
+
+        try {
+            URI pagerDutyURI = new URI(pagerDutyEndpoint);
+
+            final ActorSystem actorSystem = injector.getInstance(ActorSystem.class);
+            final ObjectMapper mapper = injector.getInstance(ObjectMapper.class);
+            final Http http = Http.get(actorSystem);
+            final ObjectNode body = mapper.createObjectNode();
+            final String subject = String.format("Alert '%s' on %s in alarm ", alert.getName(), getGroupByString(trigger));
+            final String baseUrl = typesafeConfig.getString("alerts.baseUrl");
+            final String alertUrl = URI.create(baseUrl).resolve("/#alert/edit/" + alert.getId()).toString();
+
+            body.put("service_key", pagerDutyServiceKey);
```

Comment:
Can you please move this to POJOs that we can serialize directly with the ObjectMapper?  It helps us understand how refactorings affect things and we'll get nice compile time validation.

Reply:
Will do.

Reply:
General question: when I started this work, I based it off of the 'mql' branch. Was that a mistake? Should I be looking to integrate my work into master?

Reply:
Keep working on the 'mql' branch for now.  I'm working on getting things from the 'mql' branch back into master, but the work is going a bit slower than I would like.  Master currently doesn't have any of the alerting at all.  When you get the code merged into the 'mql' branch, I'll get it shepherded into master as well.

Reply:
Thanks. I tried running the master branch yesterday - I could get it running (sort of) under docker, but had no real success running it with the other mechanisms listed in the README. There's no urgent business driver for us to be running the master version but it would be nice to have that option.